test/unity/toolkits/drawing_classifier/test_dc_prediction.cxx

```diff
@@ -0,0 +1,584 @@
+/* Copyright © 2019 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#define BOOST_TEST_MODULE test_dc_data_iterator
+
+#include <toolkits/drawing_classifier/dc_data_iterator.hpp>
+#include <toolkits/drawing_classifier/drawing_classifier.hpp>
+
+#include <algorithm>
+
+#include <boost/test/unit_test.hpp>
+#include <core/util/test_macros.hpp>
+#include <model_server/lib/image_util.hpp>
+
+#include "../neural_net/neural_net_mocks.hpp"
+#include "dc_data_utils.hpp"
+#include "dc_mock_interface.hpp"
+
+namespace turi {
+namespace drawing_classifier {
+
+namespace {
+
+using turi::neural_net::compute_context;
+using turi::neural_net::float_array_map;
+using turi::neural_net::shared_float_array;
+using turi::neural_net::mock_compute_context;
+using turi::neural_net::mock_model_backend;
+using turi::neural_net::model_backend;
+using turi::neural_net::model_spec;
+using turi::neural_net::shared_float_array;
+
+
+/* ==================== global variables  ====================== */
+
+/**
+ * tuple of batch_size, num_of_rows, num_of_classes
+ *
+ * batch_size, num_of_rows, num_of_classes
+ * 1. batch_size > num_of_rows
+ * 2. batch_size == num_of_rows
+ * 3. num_of_rows can divide batch_size
+ * 4. num_of_rows can not divide batch_size
+ *
+ **/
+std::vector<std::tuple<unsigned, unsigned, unsigned>> TEST_CASES = {
+    {2, 1, 1}, {2, 1, 2}, {2, 2, 2}, {2, 4, 2},
+    {2, 4, 3}, {2, 5, 2}, {2, 5, 3}};
+
+constexpr auto PRED_NAME = "preds";
+
+/* ============== utils borrowed from gl_sframe.cxx ============= */
+// TODO: create a common util class
+void _assert_flexvec_equals(const std::vector<flexible_type>& sa,
+                            const std::vector<flexible_type>& sb) {
+  TS_ASSERT_EQUALS(sa.size(), sb.size());
+  for (size_t i = 0; i < sa.size(); ++i) {
+    if (sa[i] != sb[i]) {
+      std::cout << sa[i] << " vs " << sb[i] << std::endl;
+    }
+    TS_ASSERT_EQUALS(sa[i], sb[i]);
+  }
+}
+
+void _assert_sframe_equals(gl_sframe sa, gl_sframe sb) {
+  TS_ASSERT_EQUALS(sa.size(), sb.size());
+  TS_ASSERT_EQUALS(sa.num_columns(), sb.num_columns());
+  auto a_cols = sa.column_names();
+  auto b_cols = sb.column_names();
+  std::sort(a_cols.begin(), a_cols.end());
+  std::sort(b_cols.begin(), b_cols.end());
+  for (size_t i = 0; i < a_cols.size(); ++i)
+    TS_ASSERT_EQUALS(a_cols[i], b_cols[i]);
+  sb = sb[sa.column_names()];
+  for (size_t i = 0; i < sa.size(); ++i) {
+    _assert_flexvec_equals(sa[i], sb[i]);
+  }
+}
```

Comment:
Can we call these functions directly by separating these utilities instead of copy-pasting code?

Reply:
see TODO